pkg/util/dynamic-resource-allocation/controller/controller.go

```diff
@@ -0,0 +1,809 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+
+	v1 "k8s.io/api/core/v1"
+	resourcev1alpha1 "k8s.io/api/resource/v1alpha1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1types "k8s.io/client-go/kubernetes/typed/core/v1"
+	resourcev1alpha1listers "k8s.io/client-go/listers/resource/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/util/dynamic-resource-allocation/resourceclaim"
+)
+
+// Controller watches ResourceClaims and triggers allocation and deallocation
+// as needed.
+type Controller interface {
```

Comment:
I don't understand why this is under util?  Why is this not all pkg/controller?  It's really really not util.

Reply:
This is not for kube-controller-manager. It's the control loop that runs inside of a resource driver.

Right now it is used by test/e2e/dra/test-driver (in-tree) and the NVIDIA driver (out-of-tree). It was meant to get published as a staging repo (k8s.io/dynamic-resource-allocation, [mail](https://groups.google.com/g/kubernetes-sig-architecture/c/DkKQyuUrj9Y) and [issue](https://github.com/kubernetes/org/issues/3821)) but there are [concerns](https://kubernetes.slack.com/archives/C01672LSZL0/p1667751787905519) about problems with the release process when creating a new staging repo now.

With that explanation in mind and considering that `pkg/util` is only a stop-gap solution for one release, is it the right place or should I move it elsewhere? Where?

Reply:
We can not be telling people to vendor k/k - it's a nightmare.

If this is O(days) while we get the rest set up, OK

Reply:
if the feature is alpha, doesn't it make sense to have the code somewhere in k/k?

Unless we have very clear messaging in the staging repo that it has zero stability guarantees (or that it can even be deleted).

Reply:
@mrbobbytables is ready to create the staging repo for this code. He is just waiting for the exception request to be approved and for this PR to get closer to merging. A safer approach would be to first merge this PR, then set up the staging repo, and then do a PR which moves the code back.

I think we can have a staging repo with an even stronger warning than usual ("alpha level code", "may get deleted").

Reply:
The warning in the README for the staging repo was (and will be again, if I revert):

> ## Compatibility
> 
> There are *NO compatibility guarantees* for this repository, yet.  It is in direct support of Kubernetes, so branches
> will track Kubernetes and be compatible with that repo.

This is the standard boilerplate for a staging repo.